File: Data Structures & Algorithms/jump-game/submission-6.java
class Solution {
    public boolean canJump(int[] nums) {
        int n = nums.length;
        int goal = nums.length - 1;
        for (int i = n - 1; i >= 0; i--) {
            if (nums[i] + i >= goal) {
                goal = i;
            }
        }
        return goal == 0;
    }
}

File: Data Structures & Algorithms/search-2d-matrix/submission-6.py
def binary_search(nums: List[int], target: int) -> int:
    left = 0
    right = len(nums) - 1
    while left <= right:
        mid = (left + right) // 2
        if nums[mid] < target:
            left = mid + 1
        elif nums[mid] > target:
            right = mid - 1
        elif nums[mid] == target:
            return mid
    return -1

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        for m in matrix:
            index = binary_search(m, target)
            if (index != -1): 
                return True
        return False